game_core/types.proto

```diff
@@ -0,0 +1,144 @@
+syntax = "proto3";
+
+package Core;
+
+// Args
+// client
+message LoginArgs {
+  string userName = 1;
+  string password = 2;
+}
+
+message QueryAccountArgs {
+  oneof key {
+    string userName = 1;
+    // could get private info
+    string token = 2;
+  }
+}
+
+// room
+message CreateRoomArgs {
+  string gameName = 1;
+  string roomPassword = 2;
+
+  google.protobuf.Any custom = 11;
+}
+
+message ListRoomArgs {
+  string gameName = 1;
+}
+
+message JoinRoomArgs {
+  string token = 1;
+  int32 roomId = 2;
+}
+
+message ExitRoomArgs {
+  string token = 1;
+}
+
+message QueryRoomArgs {
+  string token = 1;
+  int32 roomId = 2;
+}
+
+// ready, start, playX
+message OperationInRoomArgs {
+  string token = 1;
+  int32 roomId = 2;
+  oneof RoomOperation {
+    MessageArgs messageArgs = 1;
+    // room owner use this to start game
+    bool ready = 2;
+
+    google.protobuf.Any custom = 11;
+  }
+}
+
+// chat
+message MessageArgs {
+  string token = 1;
+  string message = 2;
+}
+
+// broadcast
+message ListenArgs {
+  int32 sequenceId = 1;
+
+  oneof broadcast {
+    // chat
+    string message = 1;
+
+    // room operation
+    // notify the change of a room to the user in that room
+    JoinRoomArgs joinRoomArgs = 2;
+    ExitRoomArgs exitRoomArgs = 3;
+
+    // game operation
+    google.protobuf.Any custom = 11;
+  }
+}
+
+// Ret
+// client
+message LoginRet {
+  ErrorNumber errno = 1;
+  string token = 2;
+  Account account = 3;
+}
+
+message QueryAccountRet {
+  ErrorNumber errno = 1;
+  Account account = 2;
+}
+
+// room
+message CreateRoomRet {
+  ErrorNumber errno = 1;
+}
+message ListRoomRet {
+  ErrorNumber errno = 1;
+  repeated Room room = 2;
+}
+message JoinRoomRet {
+  ErrorNumber errno = 1;
+}
+message ExitRoomRet {
+  ErrorNumber errno = 1;
+}
+message QueryRoomRet {
+  ErrorNumber errno = 1;
+}
+
+// ready, start, playX
+message OperationInRoomRet {
+  ErrorNumber errno = 1;
+}
+
+// chat
+message MessageRet {
+  ErrorNumber errno = 1;
+}
+
+// broadcast
+message ListenRet {
+  ErrorNumber errno = 1;
+}
+
+// Error number
+enum ErrorNumber {
+  OK = 0;
+  UnkownError = 1;
+}
+
+message Account {
+  string userName = 1;
+  int32 points = 2;
+}
+
+message Room {
+  string gameName = 1;
+  int32 roomId = 2;
+  repeated string userName = 3;
```

Comment:
room should have the meta data about the game specific, like the num of member limit

Reply:
indeed,  need more discussion, if we just want to expose the game setting to client, how about use a DSL?
e.g. the game defines:  
{
'EnableX': ['bool', 'true', 'Description', 'validation rule']   
'MaxUser': ['int32', '6', 'Desp', 'validation rule: ']
}